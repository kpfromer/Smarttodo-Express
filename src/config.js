import convict from 'convict';

// Define a schema
const config = convict({
  env: {
    doc: "The application environment.",
    format: ["production", "development", "test"],
    default: "development",
    env: "NODE_ENV"
  },
  port: {
    doc: "The port for express to bind.",
    format: "port",
    default: 3001,
    env: "PORT",
    arg: "port"
  },
  jwt: {
    secret: {
      doc: "The jwt secret used to encode tokens",
      format: String,
      default: "jwt-dev-secret",
      env: "JWT_SECRET"
    },
    expireIn: {
      doc: "The time for the token to expire in",
      format: [String, Number],
      default: "1h",
      env: "JWT_EXPIRE"
    }
  },
  db: {
    url: {
      doc: "The mongo url used for connection.",
      format: "*", // TODO: validate for mongo url
      default: "mongodb://localhost:27017/smarttodo-express",
      env: "MONGO_URL"
    }
  }
});

export default config;