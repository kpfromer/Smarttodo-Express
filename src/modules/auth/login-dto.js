import joi from 'joi';

export const loginDto = joi.object().keys({
  username: joi.string().required(),
  password: joi.string().required()
}).required();