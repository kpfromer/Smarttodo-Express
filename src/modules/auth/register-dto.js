import joi from 'joi';

export const registerDto = joi.object().keys({
  username: joi.string().required().min(4).max(30),
  password: joi.string().required().min(3).max(60),
  email: joi.string().required().email(),
  firstName: joi.string().required().trim().min(2).max(30).regex(/^[a-z]+$/i, { name: 'alphabet' }),
  lastName: joi.string().required().trim().min(2).max(30).regex(/^[a-z]+$/i, { name: 'alphabet' })
}).required();