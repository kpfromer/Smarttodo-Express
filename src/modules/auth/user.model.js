import mongoose from 'mongoose';
import bcrypt from 'bcryptjs';
import { promisify } from 'util';

const schema = new mongoose.Schema({
  username: { required: true, type: String, unique: true },
  password: { required: true, type: String },
  email: { required: true, type: String, unique: true },
  firstName: { required: true, type: String },
  lastName: { required: true, type: String }
});

schema.pre('save', async function (next) {
  const encryptedPassword = await promisify(bcrypt.hash)(this.password, 8);
  this.password = encryptedPassword;
  next();
});

schema.methods.checkPassword = async function(plainTextPassword) {
  return await bcrypt.compare(plainTextPassword, this.password);
}

const model = mongoose.model('User', schema);

export default model;