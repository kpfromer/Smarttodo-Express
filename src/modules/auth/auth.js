import boom from 'boom';
import User from './user.model';
import jwt from 'jsonwebtoken';
import config from '../../config';

export const registerUser = async userDetails => {
  const { username, password, email, firstName, lastName } = userDetails;
  const user = await User.findOne({ $or: [{username}, {email}] }).exec();

  if (!!user) {
    let message = 'Username already taken';
    if (user.username === username && user.email === email) {
      message = 'Username/email already taken';
    } else if (user.email === email) {
      message = 'Email already taken';
    }
    
    throw boom.badRequest(message);
  }

  await User.create({
    username,
    password,
    email,
    firstName,
    lastName
  })

  return {
    success: true,
    message: 'Successfully registered'
  };
}

export const loginUser = async userDetails => {
  const { username, password } = userDetails;

  const user = await User.findOne({ username }).exec();
  if (
    !!user &&
    user.checkPassword(password)
  ) {
    return {
      access_token: jwt.sign({ id: user._id }, config.get('jwt.secret'), { expiresIn: config.get('jwt.expireIn') })
    };
  }
  throw boom.badRequest('Username/password do not match');
}