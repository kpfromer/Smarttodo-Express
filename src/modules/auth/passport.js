import { Strategy as JwtStrategy, ExtractJwt } from 'passport-jwt';
import User from './user.model';
import config from '../../config';

export const findUser = async jwtPayload => {
  const user = await User.findOne({ _id: jwtPayload.id }).exec();
  if (!!user) {
    return user;
  }
  return false;
}

export const errorCatcher = asyncFn => (jwtPayload, done) =>
  asyncFn(jwtPayload)
    .then(authenticated => done(null, authenticated))
    .catch(error => done(error, false));

const generateJwtStrategy = () => {
  const options = {
    secretOrKey: config.get('jwt.secret'),
    jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken()
  };
  return new JwtStrategy(options, errorCatcher(findUser));
}

export default generateJwtStrategy;