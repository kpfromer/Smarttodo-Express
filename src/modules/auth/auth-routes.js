import express from 'express';
import { registerUser, loginUser } from './auth';
import validate from '../../helper/validate';
import asyncMiddleware from '../../helper/async-middleware';
import { loginDto } from './login-dto';
import { registerDto } from './register-dto';

const router = express.Router();

router.post('/register', validate(registerDto), asyncMiddleware(async (req, res) =>
  res.json(await registerUser(req.validatedBody))
));

router.post('/login', validate(loginDto), asyncMiddleware(async (req, res) =>
  res.json(await loginUser(req.validatedBody))
));

export default router;