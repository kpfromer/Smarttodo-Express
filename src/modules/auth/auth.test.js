import User from './user.model';
import { registerUser, loginUser } from './auth';
import jwt from 'jsonwebtoken';
import config from '../../config';

// Setup mock model
jest.mock('./user.model', () => ({
  findOne: jest.fn(),
  create: jest.fn()
}));

// Mock jwt sign
jest.mock('jsonwebtoken', () => ({
  sign: jest.fn()
}));

// Mock config (jwt)
jest.mock('../../config', () => ({
  get: jest.fn()
}))

describe('auth', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });
  describe('registerUser', () => {
    let exec;
    beforeEach(() => {
      exec = jest.fn();
      User.findOne.mockReturnValue({
        exec
      });
    });
    
    it('creates a user', async () => {
      exec.mockResolvedValue(null);
      // jest.spyOn(User.findOne, 'exec').mockResolvedValue(null);
      const userDetails = {
        username: 'hello',
        password: 'whatupworld',
        firstName: 'jack',
        lastName: 'Myers',
        email: 'jmyers@example.com'
      };

      const returnValue = await registerUser(userDetails)

      // expect(res.status).toBe(200);
      expect(User.findOne).toHaveBeenCalledWith({ $or: [{ username: userDetails.username }, { email: userDetails.email }] })
      expect(exec).toHaveBeenCalled();
      expect(User.create).toBeCalledWith(userDetails);
      expect(returnValue).toMatchSnapshot();
    });
    it('throws an error when username is already taken', async () => {
      exec.mockResolvedValue({
        username: 'hello',
        password: 'adsfasd',
        firstName: 'mick',
        lastName: 'jackson',
        email: 'mjaskson@example.com'
      });
      let thrownError;
      // jest.spyOn(User.findOne, 'exec').mockResolvedValue(null);
      const userDetails = {
        username: 'hello',
        password: 'whatupworld',
        firstName: 'jack',
        lastName: 'Myers',
        email: 'jmyers@example.com'
      };

      try {
        await registerUser(userDetails)
      } catch (error) {
        thrownError = error;
      }

      expect(thrownError).toMatchSnapshot();
      expect(User.create).not.toBeCalled();
    });
    it('throws an error when email is already taken', async () => {
      exec.mockResolvedValue({
        username: 'adfasdfasdfhadsf',
        password: 'adsfasd',
        firstName: 'mick',
        lastName: 'jackson',
        email: 'jmyers@example.com'
      });
      let thrownError;
      // jest.spyOn(User.findOne, 'exec').mockResolvedValue(null);
      const userDetails = {
        username: 'hello',
        password: 'whatupworld',
        firstName: 'jack',
        lastName: 'Myers',
        email: 'jmyers@example.com'
      };

      try {
        await registerUser(userDetails)
      } catch (error) {
        thrownError = error;
      }

      expect(thrownError).toMatchSnapshot();
      expect(User.create).not.toBeCalled();
    });
  }); 
  describe('loginUser', () => {
    let exec, checkPassword;
    beforeEach(() => {
      exec = jest.fn();
      checkPassword = jest.fn()
      User.findOne.mockReturnValue({
        exec
      });
      exec.mockResolvedValue({
        _id: 'the-id-of-the-user',
        checkPassword
      });
    });
    it('logins in the user', async () => {
      config.get.mockImplementation(value => {
        if (value === 'jwt.secret')
          return 'JWT-SECRET-TESTING'
        else if (value === 'jwt.expireIn')
          return 'JWT-EXPIRE-IN';
        else
          throw new Error('Unknown config get value! (TESTING)')
      });
      jwt.sign.mockReturnValue('ENCRYPTED-JWT-TOKEN');
      checkPassword.mockReturnValue(true);
      
      const returnValue = await loginUser({ username: 'jackson', password: 'whatupworld!'})
      expect(User.findOne).toHaveBeenCalledWith({ username: 'jackson' });
      expect(checkPassword).toHaveBeenCalledWith('whatupworld!');
      expect(jwt.sign.mock.calls[0]).toMatchSnapshot();
      expect(returnValue).toMatchSnapshot();
    });
    it('errors if passwords don\'t match', async () => {
      let thrownError;
      checkPassword.mockReturnValue(false);
      
      try {
        await loginUser({ username: 'jackson', password: 'whatupworld!'})
      } catch(error) {
        thrownError = error;
      }
      expect(checkPassword).toHaveBeenCalledWith('whatupworld!');
      expect(thrownError).toMatchSnapshot();
    });
    it('errors if the user doesn\'t exist', async () => {
      let thrownError;
      exec.mockReturnValue(null);
      
      try {
        await loginUser({ username: 'jackson', password: 'whatupworld!'})
      } catch(error) {
        thrownError = error;
      }
      expect(thrownError).toMatchSnapshot();
    });
  });
})