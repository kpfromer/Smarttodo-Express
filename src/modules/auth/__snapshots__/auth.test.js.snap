// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth loginUser errors if passwords don't match 1`] = `[Error: Username/password do not match]`;

exports[`auth loginUser errors if the user doesn't exist 1`] = `[Error: Username/password do not match]`;

exports[`auth loginUser logins in the user 1`] = `
Array [
  Object {
    "id": "the-id-of-the-user",
  },
  "JWT-SECRET-TESTING",
  Object {
    "expiresIn": "JWT-EXPIRE-IN",
  },
]
`;

exports[`auth loginUser logins in the user 2`] = `
Object {
  "access_token": "ENCRYPTED-JWT-TOKEN",
}
`;

exports[`auth registerUser creates a user 1`] = `
Object {
  "message": "Successfully registered",
  "success": true,
}
`;

exports[`auth registerUser throws an error when email is already taken 1`] = `[Error: Email already taken]`;

exports[`auth registerUser throws an error when username is already taken 1`] = `[Error: Username already taken]`;
